crates/api/src/chronicle_graphql/mod.rs

```diff
@@ -373,37 +374,134 @@ where
+    QUERY,
+    MUTATE,
+    SUBSCRIBE,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct AuthGuard {
+    access_type: AccessType,
+    method_name: String,
+}
+
+impl AuthGuard {
+    pub fn new(access_type: AccessType, method_name: impl ToString) -> Self {
+        AuthGuard {
+            access_type,
+            method_name: method_name.to_string(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl async_graphql::Guard for AuthGuard {
+    async fn check(&self, ctx: &Context<'_>) -> async_graphql::Result<()> {
+        if let (Some(identity), Some(opa_executor)) = (
+            ctx.data_opt::<AuthId>(),
+            ctx.data_opt::<Arc<Mutex<WasmtimeOpaExecutor>>>(),
+        ) {
+            let context = serde_json::to_value(self)?;
+            if opa_executor
+                .lock()
+                .await
+                .evaluate(identity, &context)
+                .await?
+            {
+                Ok(())
+            } else {
+                Err("violation of policy rules".into())
+            }
+        } else {
+            Err("cannot check policy rules".into())
+        }
+    }
+}
+
 #[derive(Clone, Debug)]
 pub struct JwtClaims(pub serde_json::Map<String, serde_json::Value>);
 
-struct AuthorizationEndpoint<Q, M, S> {
+struct AuthorizationEndpointQuery<Q, M, S> {
     checker: JwtChecker,
     schema: Schema<Q, M, S>,
 }
 
 #[poem::async_trait]
-impl<Q, M, S> Endpoint for AuthorizationEndpoint<Q, M, S>
+impl<Q, M, S> Endpoint for AuthorizationEndpointQuery<Q, M, S>
 where
     Q: ObjectType + 'static,
     M: ObjectType + 'static,
     S: SubscriptionType + 'static,
 {
-    type Output = GraphQLBatchResponse;
+    type Output = poem::Response;
 
     async fn call(&self, req: poem::Request) -> poem::Result<Self::Output> {
         if let Some(authorization) = req.header("Authorization") {
             if let Ok(authorization) = HeaderValue::from_str(authorization) {
                 let bearer_token_maybe: Option<Bearer> = Credentials::decode(&authorization);
                 if let Some(bearer_token) = bearer_token_maybe {
                     if let Ok(claims) = self.checker.verify_jwt(bearer_token.token()).await {
-                        use poem::FromRequest;
+                        use poem::{FromRequest, IntoResponse};
                         let (req, mut body) = req.split();
                         let req = GraphQLBatchRequest::from_request(&req, &mut body).await?;
                         let req = req.0.data(JwtClaims(claims));
                         let span = trace_span!("Execute GraphQL request", request = ?req);
                         return Ok(GraphQLBatchResponse(
                             self.schema.execute_batch(req).instrument(span).await,
-                        ));
+                        )
+                        .into_response());
+                    }
+                }
+            }
+            tracing::warn!("rejected authorization: {:?}", authorization);
```

Comment:
can we use the non string form for events, i.e (rejected_authorisation = ?authorisation), this will be much easier to query for in log aggregators than string formatting. Remind me when I forget this too!

Reply:
That `authorization` is just the string from the HTTP header but I'll add a log line for if the credentials got decoded.

Reply:
Ah no, it's the message I am referring to.

```
"rejected authorization: {:?}", authorization
```

Will come up as "rejected authorisation {literal json}" in a string

rejected_authorisation=?authorisation will emit structure log data, not a string that needs to be parsed

Reply:
The `authorization` string just has `Bearer base64stuff....` - but I can try the `?` approach on the `bearer_token` struct if we manage to parse the base64stuff out of it?

Reply:
It doesn't look as if the `bearer_token` has JSON in it either. Well, it was worth a look.